tests/integration_tests/bugs/test_lp1835584.py

```diff
@@ -0,0 +1,82 @@
+""" Integration test for LP #1835584
+
+Upstream linux kernels prior to 4.15 providate DMI product_uuid in uppercase.
+More recent kernels switched to lowercase for DMI product_uuid. Azure
+datasource uses this product_uuid as the instance-id for cloud-init.
+
+The linux-azure-fips kernel installed in PRO FIPs images, that product UUID is
+uppercase whereas the linux-azure cloud-optimized kernel reports the UUID as
+lowercase.
+
+In cases where product_uuid changes case, ensure cloud-init doesn't
+recreate ssh hostkeys across reboot (due to detecting an instance_id change).
+
+This currently only affects linux-azure-fips -> linux-azure on Bionic.
+This test won't run on Xenial because both linux-azure-fips and linux-azure
+report uppercase product_uuids.
+
+The test will launch a Bionic Ubuntu PRO FIPS image which has
+linux-azure-fips kernel and reports product_uuid as uppercase. Then upgrade
+and reboot into linux-azure kernel which is known to report product_uuid as
+lowercase.
+
+Across the reboot, assert that we didn't re-run config_ssh by virtue of
+seeing only one semaphore creation log entry of type:
+
+ Writing to /var/lib/cloud/instances/<UUID>/sem/config_ssh -
+
+https://bugs.launchpad.net/cloud-init/+bug/1835584
+"""
+import re
+
+import pytest
+
+from tests.integration_tests.clouds import IntegrationCloud
+
+
+IMG_AZURE_UBUNTU_PRO_FIPS_BIONIC = (
+    "Canonical:0001-com-ubuntu-pro-bionic-fips:pro-fips-18_04:18.04.202010201"
+)
+
+
+def _check_iid_insensitive_across_kernel_upgrade(instance):
+    uuid = instance.read_from_file('/sys/class/dmi/id/product_uuid')
+    assert uuid.isupper(), (
+        "Expected uppercase UUID on Ubuntu FIPS image {}".format(
+            uuid
+        )
+    )
+    orig_kernel = instance.execute('uname -r').strip()
+    assert "azure-fips" in orig_kernel
+    result = instance.execute("apt-get update")
+    # Install a 5.4+ kernel which provides lowercase product_uuid
+    result = instance.execute("apt-get install linux-azure")
+    if not result.ok:
+        pytest.fail(
+            "Unable to install linux-azure kernel: {}".format(result)
+        )
+    instance.restart()
+    new_kernel = instance.execute('uname -r').strip()
+    assert orig_kernel != new_kernel
+    assert 'azure-fips' not in new_kernel
+    assert 'azure' in new_kernel
+    new_uuid = instance.read_from_file('/sys/class/dmi/id/product_uuid')
+    assert uuid.lower() == new_uuid, (
+        "Expected UUID on linux-azure to be lowercase of FIPS: {}".format(uuid)
+    )
+    log = instance.read_from_file('/var/log/cloud-init.log')
+    RE_CONFIG_SSH_SEMAPHORE = r'Writing.*sem/config_ssh '
+    ssh_runs = len(re.findall(RE_CONFIG_SSH_SEMAPHORE, log))
+    assert 1 == ssh_runs, "config_ssh ran too many times {}".format(ssh_runs)
+
+
+@pytest.mark.azure
+@pytest.mark.sru_next
+@pytest.mark.not_xenial
```

Comment:
This will still run on every release bionic+; we *could* introduce an `only_bionic` mark or similar. However, given that this test doesn't use the `client` fixtures, we don't actually need to perform the skip via marks to avoid the instance launch: we can do it in the body of the test before we call `launch` instead.

Ideally, we'd be able to do something like `if (session_cloud.image.os, session_cloud.image.release) != ("ubuntu", "bionic"): pytest.skip(...)` but we don't currently store the `ImageSpecification` that we determine on `session_cloud`. That's probably not too painful to hook up (though may require some thought around snapshotting), but I don't think we need to for this, re-parsing `OS_IMAGE` should be sufficient: something like `image_spec = ImageSpecification.from_os_image(); if (image_spec.os, image_spec.release) != ("ubuntu", "bionic"): pytest.skip(...)`.

Reply:
@OddBloke funny, I put that type of pytest.skip in locally and decided to pullt it out based on looking at self.settings.os_image I think. I'll put some semblance of that back into this PR